File: src/test/java/com/jvanev/jconfig/converter/internal/ValueConverterTest.java
/*
 * Copyright 2025 Georgi Vanev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jvanev.jconfig.converter.internal;

import com.jvanev.jconfig.exception.ValueConversionException;
import java.time.LocalDateTime;
import java.time.temporal.Temporal;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertIterableEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class ValueConverterTest {
    private final ValueConverter converter = new ValueConverter();

    @Test
    void shouldConvertPrimitives() {
        var byteType = "127";
        var shortType = "16325";
        var intType = "1000001";
        var longType = "1000000001";
        var floatType = "1.23";
        var doubleType = "1.23";
        var booleanType = "True";
        var charType = "a";

        assertEquals((byte) 127, converter.convert(byte.class, byteType));
        assertEquals((short) 16325, converter.convert(short.class, shortType));
        assertEquals(1_000_001, converter.convert(int.class, intType));
        assertEquals(1_000_000_001L, converter.convert(long.class, longType));
        assertEquals(1.23f, converter.convert(float.class, floatType));
        assertEquals(1.23, converter.convert(double.class, doubleType));
        assertEquals(true, converter.convert(boolean.class, booleanType));
        assertEquals('a', converter.convert(char.class, charType));
    }

    @Test
    void shouldConvertHexStringToNumeric() {
        var byteType = "0x07";
        var shortType = "0x08";
        var intType = "0x09";
        var longType = "0x0A";

        assertEquals((byte) 7, converter.convert(byte.class, byteType));
        assertEquals((short) 8, converter.convert(short.class, shortType));
        assertEquals(9, converter.convert(int.class, intType));
        assertEquals(10L, converter.convert(long.class, longType));
    }

    @Test
    void shouldConvertToPrimitiveArrays() {
        var intValues = "1, 2, 0x03, #4";
        var expectedIntArray = new int[]{1, 2, 3, 4};
        assertArrayEquals(expectedIntArray, (int[]) converter.convert(int[].class, intValues));

        var booleanValues = "True, false, TRUE , FALSe";
        var expectedBooleanArray = new boolean[]{true, false, true, false};
        assertArrayEquals(
            expectedBooleanArray,
            (boolean[]) converter.convert(boolean[].class, booleanValues)
        );

        var charValues = "a, b, c,d";
        var expectedCharArray = new char[]{'a', 'b', 'c', 'd'};
        assertArrayEquals(expectedCharArray, (char[]) converter.convert(char[].class, charValues));

        var byteValues = "1, 2, 3";
        var expectedByteArray = new byte[]{1, 2, 3};
        assertArrayEquals(expectedByteArray, (byte[]) converter.convert(byte[].class, byteValues));

        var shortValues = "100, 200, 300";
        var expectedShortArray = new short[]{100, 200, 300};
        assertArrayEquals(expectedShortArray, (short[]) converter.convert(short[].class, shortValues));

        var longValues = "10000000000, 20000000000";
        var expectedLongArray = new long[]{10000000000L, 20000000000L};
        assertArrayEquals(expectedLongArray, (long[]) converter.convert(long[].class, longValues));

        var floatValues = "1.1, 2.2, 3.3";
        var expectedFloatArray = new float[]{1.1f, 2.2f, 3.3f};
        assertArrayEquals(expectedFloatArray, (float[]) converter.convert(float[].class, floatValues));

        var doubleValues = "1.1, 2.2, 3.3";
        var expectedDoubleArray = new double[]{1.1, 2.2, 3.3};
        assertArrayEquals(expectedDoubleArray, (double[]) converter.convert(double[].class, doubleValues));
    }

    @Test
    void shouldThrowOnInvalidCharArrayElement() {
        var invalidCharValues = "a, bc, d"; // 'bc' is not a single character

        assertThrows(
            IllegalArgumentException.class,
            () -> converter.convert(char[].class, invalidCharValues)
        );
    }

    static class ExampleType {
        List<Integer> list;
        Map<String, Integer> map;
        Stack<Integer> stack;
    }

    static class Stack<T> {
        private final List<T> items = new LinkedList<>();

        int getSize() {
            return items.size();
        }

        void push(T item) {
            items.add(item);
        }

        T pop() {
            return items.remove(items.size() - 1);
        }

        List<T> all() {
            return items.stream().toList();
        }
    }

    @Test
    void shouldConvertToCollection() throws NoSuchFieldException {
        var entries = "1, 2, 3, 4";
        var type = ExampleType.class.getDeclaredField("list").getGenericType();

        assertIterableEquals(List.of(1, 2, 3, 4), (List<?>) converter.convert(type, entries));
    }

    @Test
    void shouldConvertToMap() throws NoSuchFieldException {
        var entries = "Skill1: 7200, Skill2 :3600, Skill3 : 1800, Skill4:900";
        var type = ExampleType.class.getDeclaredField("map").getGenericType();
        var expectedResult = new LinkedHashMap<String, Integer>();

        expectedResult.put("Skill1", 7200);
        expectedResult.put("Skill2", 3600);
        expectedResult.put("Skill3", 1800);
        expectedResult.put("Skill4", 900);

        assertEquals(expectedResult, converter.convert(type, entries));
        assertEquals(Map.of(), converter.convert(type, ""));
    }

    @Test
    void shouldThrowOnMalformedMapEntries() throws NoSuchFieldException {
        var type = ExampleType.class.getDeclaredField("map").getGenericType();

        assertThrows(IllegalArgumentException.class, () -> converter.convert(type, "Skill1 ; 7200"));
        assertThrows(IllegalArgumentException.class, () -> converter.convert(type, "Skill1, : 7200"));
    }

    @Test
    void shouldUseCustomConvertorWithAssignableReturnValue_IfNoDirectTypeMatchFound() {
        var converter = new ValueConverter();
        converter.addValueConverter(
            Temporal.class,
            (type, typeArguments, value) -> LocalDateTime.parse(value)
        );

        var value = "2025-07-31T17:55:12";

        assertEquals(LocalDateTime.parse(value), converter.convert(LocalDateTime.class, value));
    }

    @Test
    void shouldSupportCustomGenericTypes() throws NoSuchFieldException {
        var entries = "1, 2, 3, 4";
        var converter = new ValueConverter();
        converter.
            addValueConverter(
                Stack.class,
                (type, typeArguments, value) -> {
                    var array = value.split("\\s*,\\s*");
                    var stack = new Stack<>();

                    for (var item : array) {
                        stack.push(converter.convert(typeArguments[0], item));
                    }

                    return stack;
                }
            );

        var type = ExampleType.class.getDeclaredField("stack").getGenericType();
        var stack = (Stack<?>) converter.convert(type, entries);

        assertEquals(List.of(1, 2, 3, 4), stack.all());
    }

    enum LogLevel {
        DEBUG, INFO
    }

    @Test
    void shouldConvertEnums() {
        var debug = "DEBUG";
        var info = "INFO";

        assertEquals(LogLevel.DEBUG, converter.convert(LogLevel.class, debug));
        assertEquals(LogLevel.INFO, converter.convert(LogLevel.class, info));
    }

    public record TestType(int value) {
        public static TestType valueOf(String value) {
            return new TestType(Integer.parseInt(value));
        }
    }

    @Test
    void shouldConvertTypesWithValueOfMethod() {
        var value = "127";

        assertEquals(new TestType(127), converter.convert(TestType.class, value));
    }

    public record TestTypeWithThrowingValueOf(int value) {
        public static TestTypeWithThrowingValueOf valueOf(String value) {
            throw new RuntimeException();
        }
    }

    @Test
    void shouldThrowOnValueOfInvocationException() {
        assertThrows(
            ValueConversionException.class,
            () -> converter.convert(TestTypeWithThrowingValueOf.class, "test")
        );
    }
}

File: src/test/kotlin/converter/ValueConverterTest.kt
/*
 * Copyright 2025 Georgi Vanev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jvanev.kconfig.converter

import com.jvanev.kconfig.ValueConversionException
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.time.LocalDateTime
import java.time.temporal.Temporal
import java.util.LinkedList
import kotlin.test.assertEquals

class ValueConverterTest {
    private val converter = ValueConverter()

    @Test
    fun shouldConvertPrimitives() {
        val byte = "127"
        val short = "16325"
        val int = "1000001"
        val long = "1000000001"
        val float = "1.23"
        val double = "1.23"
        val boolean = "True"
        val char = "a"

        assertEquals(127.toByte(), converter.convert(byte, Byte::class.java))
        assertEquals(16325.toShort(), converter.convert(short, Short::class.java))
        assertEquals(1_000_001, converter.convert(int, Int::class.java))
        assertEquals(1_000_000_001L, converter.convert(long, Long::class.java))
        assertEquals(1.23f, converter.convert(float, Float::class.java))
        assertEquals(1.23, converter.convert(double, Double::class.java))
        assertEquals(true, converter.convert(boolean, Boolean::class.java))
        assertEquals('a', converter.convert(char, Char::class.java))
    }

    @Test
    fun shouldConvertHexStringToNumeric() {
        val byte = "0x07"
        val short = "0x08"
        val int = "0x09"
        val long = "0x0A"

        assertEquals(7.toByte(), converter.convert(byte, Byte::class.java))
        assertEquals(8.toShort(), converter.convert(short, Short::class.java))
        assertEquals(9, converter.convert(int, Int::class.java))
        assertEquals(10L, converter.convert(long, Long::class.java))
    }

    @Test
    fun shouldConvertToPrimitiveArrays() {
        val intValues = "1, 2, 0x03, #4"
        val expectedIntArray = intArrayOf(1, 2, 3, 4)
        assertArrayEquals(expectedIntArray, converter.convert(intValues, IntArray::class.java) as IntArray)

        val booleanValues = "True, false, TRUE , FALSe"
        val expectedBooleanArray = booleanArrayOf(true, false, true, false)
        assertArrayEquals(
            expectedBooleanArray,
            converter.convert(booleanValues, BooleanArray::class.java) as BooleanArray
        )

        val charValues = "a, b, c,d"
        val expectedCharArray = charArrayOf('a', 'b', 'c', 'd')
        assertArrayEquals(expectedCharArray, converter.convert(charValues, CharArray::class.java) as CharArray)

        val byteValues = "1, 2, 3"
        val expectedByteArray = byteArrayOf(1, 2, 3)
        assertArrayEquals(expectedByteArray, converter.convert(byteValues, ByteArray::class.java) as ByteArray)

        val shortValues = "100, 200, 300"
        val expectedShortArray = shortArrayOf(100, 200, 300)
        assertArrayEquals(expectedShortArray, converter.convert(shortValues, ShortArray::class.java) as ShortArray)

        val longValues = "10000000000, 20000000000"
        val expectedLongArray = longArrayOf(10000000000L, 20000000000L)
        assertArrayEquals(expectedLongArray, converter.convert(longValues, LongArray::class.java) as LongArray)

        val floatValues = "1.1, 2.2, 3.3"
        val expectedFloatArray = floatArrayOf(1.1f, 2.2f, 3.3f)
        assertArrayEquals(expectedFloatArray, converter.convert(floatValues, FloatArray::class.java) as FloatArray)

        val doubleValues = "1.1, 2.2, 3.3"
        val expectedDoubleArray = doubleArrayOf(1.1, 2.2, 3.3)
        assertArrayEquals(expectedDoubleArray, converter.convert(doubleValues, DoubleArray::class.java) as DoubleArray)
    }

    @Test
    fun shouldThrowOnInvalidCharArrayElement() {
        val charArrayType = CharArray::class.java
        val invalidCharValues = "a, bc, d" // 'bc' is not a single character

        assertThrows<IllegalArgumentException> {
            converter.convert(invalidCharValues, charArrayType)
        }
    }

    class ExampleType(
        val list: List<Int>,
        val map: Map<String, Int>,
        val stack: Stack<Int>,
    )

    class Stack<T> {
        private val items: MutableList<T> = LinkedList()

        val size: Int get() = items.size

        fun push(item: T) {
            items.add(item)
        }

        fun pop(): T {
            return items.removeLast()
        }

        fun all(): List<T> {
            return items.toList()
        }
    }

    @Test
    fun shouldConvertToCollection() {
        val entries = "1, 2, 3, 4"
        val type = ExampleType::class.java.getDeclaredField("list").genericType

        assertEquals(arrayListOf(1, 2, 3, 4), converter.convert(entries, type))
    }

    @Test
    fun shouldConvertToMap() {
        val entries = "Skill1: 7200, Skill2 :3600, Skill3 : 1800, Skill4:900"
        val type = ExampleType::class.java.getDeclaredField("map").genericType
        val expectedResult = mapOf(
            "Skill1" to 7200,
            "Skill2" to 3600,
            "Skill3" to 1800,
            "Skill4" to 900,
        )

        assertEquals(expectedResult, converter.convert(entries, type))
        assertEquals(emptyMap<String, Int>(), converter.convert("", type))
    }

    @Test
    fun shouldThrowOnMalformedMapEntries() {
        val type = ExampleType::class.java.getDeclaredField("map").genericType

        assertThrows<IllegalArgumentException> { converter.convert("Skill1 : 7200,", type) }
        assertThrows<IllegalArgumentException> { converter.convert("Skill1 ; 7200", type) }
        assertThrows<IllegalArgumentException> { converter.convert(",Skill1 : 7200", type) }
        assertThrows<IllegalArgumentException> { converter.convert("Skill1, : 7200", type) }
    }

    @Test
    fun shouldUseCustomConvertorWithAssignableReturnValue_IfNoDirectTypeMatchFound() {
        val converter = ValueConverter().apply {
            addValueConverter(Temporal::class.java) { type, _, _ ->
                LocalDateTime.parse(type)
            }
        }
        val value = "2025-07-31T17:55:12"

        assertEquals(LocalDateTime.parse(value), converter.convert(value, LocalDateTime::class.java))
    }

    @Test
    fun shouldSupportCustomGenericTypes() {
        val entries = "1, 2, 3, 4"
        val converter = ValueConverter().apply {
            addValueConverter(Stack::class.java) { value, _, typeArgs ->
                val entries = value.split("\\s*,\\s*".toRegex())
                val stack = Stack<Any>()

                for (entry in entries) {
                    stack.push(convert(entry, typeArgs[0]))
                }

                stack
            }
        }
        val type = ExampleType::class.java.getDeclaredField("stack").genericType
        val stack = converter.convert(entries, type) as Stack<*>

        assertEquals(listOf(1, 2, 3, 4), stack.all())
    }

    enum class LogLevel {
        DEBUG, INFO
    }

    @Test
    fun shouldConvertEnums() {
        val debug = "DEBUG"
        val info = "INFO"

        assertEquals(LogLevel.DEBUG, converter.convert(debug, LogLevel::class.java))
        assertEquals(LogLevel.INFO, converter.convert(info, LogLevel::class.java))
    }

    data class TestType(val value: Int) {
        companion object {
            @JvmStatic
            fun valueOf(value: String): TestType {
                return TestType(value.toInt())
            }
        }
    }

    @Test
    fun shouldConvertTypesWithValueOfMethod() {
        val value = "127"

        assertEquals(TestType(127), converter.convert(value, TestType::class.java))
    }

    data class TestTypeWithThrowingValueOf(val value: Int) {
        companion object {
            @JvmStatic
            fun valueOf(value: String): TestTypeWithThrowingValueOf {
                throw RuntimeException()
            }
        }
    }

    @Test
    fun shouldThrowOnValueOfInvocationException() {
        assertThrows<ValueConversionException> {
            converter.convert("test", TestTypeWithThrowingValueOf::class.java)
        }
    }
}
